Use ECE buttons during style preview in settings when ECE is enabled

reykjalin opened the pull request:
Fixes #9007

## Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

- Use Stripe's Express Checkout Element (ECE) to render the preview buttons in the WooPayments settings.

### Questions for design

- ~~Should we have the preview display both buttons when available or try to always limit to just 1 button?~~ We're going to display all available buttons. See https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2211295785.
    - Safari now supports showing both Google Pay and Apple Pay so we won't always know whether the ECE will show only Google Pay or both Apple Pay or Google Pay.
    - The method used to decide whether to show only Apple Pay (via ApplePaySession.canMakePayments()) is not fool-proof. We can't tell whether a customer has a card added to their wallet, in which case the button might not show up, resulting in no button in the preview.
        - This _should_ result in an info notice informing them of what they need to do to display the buttons (see below).

| Multiple express button preview | Single express button preview |
|:---:|:---:|
| <video src="https://github.com/Automattic/woocommerce-payments/assets/13835680/d2c3dd0d-ddd4-46c4-8b46-b0c0a822f2b1"> | <video src="https://github.com/Automattic/woocommerce-payments/assets/13835680/1037e493-4340-4f19-8596-1a783c563e28"> |


**Info notice**

<img width="652" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/4bb58f17-35d9-4856-a8be-d2c3e98b0c52">


## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Setup

1. Enable ECE: `npm run wp option set _wcpay_feature_stripe_ece 1`.
2. Make sure your test store is served over public-facing, HTTPS-enabled domain. Jurassic Tube is the recommended solution.
3. (For ApplePay) Register your domain with the following request:

```sh
curl https://api.stripe.com/v1/payment_method_domains \
     # Don't forget the `:` at the end!
  -u "<secret key, sk_xxx or rk_xxx>:" \
  -H "Stripe-Account: <merchant account ID>" \
     # Example: my.domain.com
  -d domain_name="<your_url>"
```

4. Make sure WooPay is disabled.
5. Open the WooPayments settings and navigate to the Apple Pay/Google Pay customization.

### Make sure an info notice is displayed when no button can be rendered

1. Remove the HTTPS check here:

https://github.com/Automattic/woocommerce-payments/blob/d8f1ee0f87fdcedc29374419cfe8adc7415bb6be/client/settings/express-checkout-settings/payment-request-button-preview.js#L225

2. Make sure WooPay is disabled.
3. Open your site with a **non-HTTPS-enabled** domain.
4. Make sure you see the info notice that's shown above in this PR description.

### Test button settings

> [!IMPORTANT]
> You need to perform this test in both Chrome (or a Chromium based browser) logged into Google Pay and Safari with Apple Pay enabled.

1. Test the various Apple Pay/Google Pay customization settings and make sure the button preview is reactive.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
add initial version of ECE preview in settings

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9028` or branch name `add/9007-use-ece-for-express-button-preview` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/9007-use-ece-for-express-button-preview) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 23eb6927ecfe4d09ac098aff2e8adbe3b4e6cc95
- Build time: 2024-07-12 03:59:38 UTC

_Note: the build is updated when a new commit is pushed to this PR._

reykjalin committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

Commit:
refactor preview button component code organization

Commit:
fix ECE button preview width

Commit:
add notice informing merchants about the need for https when store is not served over HTTPS

Commit:
use the same notice when we fail to preview the GPay/APay buttons and when https is not detected

Commit:
add notice informing merchants when the ECE preview fails to render

reykjalin committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

Commit:
enforce no overflow and 1 column layout for ECE preview

Commit:
remove unused event handlers from ece preview

Commit:
add changelog entry

Commit:
only show Apple Pay button when preview is rendered in Safari

reykjalin commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -0,0 +1,107 @@
+
+	return (
+		<InlineNotice icon status="error" isDismissible={ false }>
+			{ __(
+				'Failed to preview the Apple Pay or Google Pay button. ' +
+					'Please ensure your store is served over HTTPS on a domain available to the public internet, ' +
+					'your device is configured to use Apple Pay or Google Pay, ' +
+					'and view this page using the Safari or Chrome browsers.',
+				'woocommerce-payments'
+			) }
```
@elizaan36 @nikkivias @pierorocca In addition to the questions in the PR description, also looking for input on the info notice here. I think it's quite verbose but couldn't find any better, simpler, or more succinct phrasing that gets the point across.

Maybe we should add this to the documentation on PRBs and link to that instead?

elizaan36 replied:
Same comment as [here](https://github.com/Automattic/woocommerce-payments/pull/9028#discussion_r1669604122) I think.

reykjalin replied:
Updated in 3e8933b17bac5d2ba7f1c67fb260cb46c85aa9ff.

reykjalin commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -67,6 +69,42 @@ const buttonSizeToPxMap = {
+	</>
+);
+
+const PreviewRequirementsNotice = () => (
+	<InlineNotice icon status="info" isDismissible={ false }>
+		{ __(
+			'To preview the express checkout buttons, ' +
+				'ensure your store is served over HTTPS on a domain available to the public internet, ' +
+				'your device is configured to use Apple Pay or Google Pay, ' +
+				'and view this page using the Safari or Chrome browsers.',
+			'woocommerce-payments'
+		) }
```
@elizaan36 @nikkivias @pierorocca In addition to the questions in the PR description, also looking for input on the info notice here. I think it's quite verbose but couldn't find any better, simpler, or more succinct phrasing that gets the point across.

Maybe we should add this to the documentation on PRBs and link to that instead?

elizaan36 replied:
hey @reykjalin You can probably use the full text but maybe break it into two sentences to make it easier to read. 

Here's an option:
> To preview the express checkout buttons, ensure your store uses HTTPS on a publicly available domain and you're viewing this page on a Safari or Chrome browser. Your device must be configured to use Apple Pay or Google Pay.

reykjalin replied:
Updated in 3e8933b17bac5d2ba7f1c67fb260cb46c85aa9ff.

reykjalin commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -0,0 +1,107 @@
+
+	const applePayType = buttonType === 'default' ? 'plain' : buttonType;
+
+	let shouldOnlyShowApplePay = false;
+	if ( window.ApplePaySession && window.ApplePaySession.canMakePayments() ) {
+		shouldOnlyShowApplePay = true;
+	}
```
I don't think this method of checking if we should show Apple Pay is fool-proof. Apple says in [their documentation](https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_js_api/checking_for_apple_pay_availability):

> <dl>
> <dt><a href="https://developer.apple.com/documentation/apple_pay_on_the_web/applepaysession/1778027-canmakepayments">canMakePayments</a></dt>
> <dd>This method verifies that the device is capable of making Apple Pay payments; it doesn’t verify that the person has a provisioned card for use with Apple Pay on the device. You can call this method at any time.</dd>
> </dl>

I think that means it's possible for `canMakePayments()` to return `true` even if the Apple Pay button can't actually be displayed?

If that happens the info notice should be triggered and let the merchant know they need to do to render the button in Safari, but it'd be better if we could just show Google Pay if they're logged into that instead?

And if we choose to always show both buttons this isn't even a problem we need to think about.

cesarcosta99 replied:
>I think that means it's possible for canMakePayments() to return true even if the Apple Pay button can't actually be displayed?

That's my understanding as well.

>And if we choose to always show both buttons this isn't even a problem we need to think about.

FWIW, I think we should be displaying both whenever possible since we're just displaying a preview of them. I saw you asked the design about that so let's wait for their decision 👀

reykjalin replied:
Displaying both as of f05a885669b79915f7e7e9f058a6b8241f28b11e per https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2211295785

reykjalin commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -0,0 +1,107 @@
+	if ( canRenderButtons ) {
+		return (
+			<div
+				key={ `${ buttonType }-${ height }-${ theme }` }
```
I had to use the `key` prop here to force a rerender of the whole component when any of these props change because I found out that the `<ExpressCheckoutElement>` isn't updated when the `options` prop changes.

This causes an unfortunate "flash" where the button disappears entirely until it can be re-rendered with the new settings. But AFAICT there is no way around it.

_Although_ I guess we could render several ECE's, each with a different configuration of the settings and the settings just dictate which ones have `display: none` set? 🤔 
That probably wouldn't cause a flash like the current implementation, but it might make loading the page a little heavy? @pierorocca @elizaan36 @nikkivias what do we think here from a design perspective? (see video in description for a demo)

reykjalin replied:
Tested this a little more and was able to remove the `height` from the key. Also didn't need to include the border radius (now that #9010 is merged) so changing the height and border radius feels much better than the theme or CTA 🎉

reykjalin replied:
Here's a demonstration of the lag in re-rendering the buttons:


https://github.com/Automattic/woocommerce-payments/assets/13835680/c5d0b279-ab80-4648-9eeb-9fe3d99c7416

elizaan36 replied:
The flash is a little odd, especially since only the Apple Pay and Google Pay buttons are flashing. 

> That probably wouldn't cause a flash like the current implementation, but it might make loading the page a little heavy?

Could you share more on how loading the page would be heavy?

reykjalin replied:
> Could you share more on how loading the page would be heavy?

Yes! The flashing is because the Stripe buttons are loaded in an `iframe` from Stripe's systems, and it takes a bit to initialize the buttons. The "solution" to the flashing I'm proposing would be to load all possible variations of the buttons, and hide them all via `display: none;`, except the one that matches the current configuration.

When the configuration is changed, instead of recreating the `iframe` with the buttons we would just hide the one that's currently visible and show the one that matches the selected configuration.

The page would be heavy to load because there would be 12 different `iframes` on the page, all showing different variations of the buttons (with only 1 of those visible at a time) instead of using just one `iframe`. Basically making the page slower to load, but it should be snappy once it's loaded.

I hope that's a good enough explanation, but please let me know if you have any further questions!

> [!IMPORTANT]
> Important disclaimer: I haven't tested the proposed solution here, but I think it should work.

reykjalin replied:
@elizaan36 @pierorocca @nikkivias just to make sure this isn't holding up merging this PR I've created https://github.com/Automattic/woocommerce-payments/issues/9091 to track this issue specifically and we can keep working on it from there, if we decide this is something we need to improve.

reykjalin replied:
FYI @bborman22 ^

reykjalin commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -101,87 +149,78 @@ const PaymentRequestButtonPreview = () => {
-									e.preventDefault();
-								} }
-								options={ {
-									paymentRequest: paymentRequest,
-									style: {
-										paymentRequestButton: {
-											type: buttonType,
-											theme: theme,
-											height: `${
-												buttonSizeToPxMap[ size ] ||
-												buttonSizeToPxMap.medium
-											}px`,
-										},
-									},
-								} }
-							/>
-						) }
-				</div>
-			) }
-			{ ! isWooPayEnabled && ! isPaymentRequestEnabled && (
-				<InlineNotice icon status="info" isDismissible={ false }>
-					{ __(
-						'To preview the express checkout buttons, ' +
-							'activate at least one express checkout.',
-						'woocommerce-payments'
-					) }
-				</InlineNotice>
-			) }
-			{ isPaymentRequestSettingsPage() &&
-				isPaymentRequestEnabled &&
-				! isLoading &&
-				! paymentRequest && (
-					<InlineNotice icon status="info" isDismissible={ false }>
-						{ __(
-							'To preview the Apple Pay and Google Pay buttons, ' +
-								'ensure your device is configured to accept Apple Pay or Google Pay, ' +
-								'and view this page using the Safari or Chrome browsers.',
-							'woocommerce-payments'
-						) }
-					</InlineNotice>
```
I refactored this whole section to make it easier to read. No more ternary-hell and instead there are several if-statements I find much easier to read. It might be easier to look at the changes here in a split diff — or even locally in your editor — rather than trying to parse this with the default view.

reykjalin commented on client/settings/wcpay-settings-context.js:
```diff
@@ -11,6 +11,7 @@ const WCPaySettingsContext = createContext( {
 		isAuthAndCaptureEnabled: false,
 		isDisputeIssuerEvidenceEnabled: false,
 		woopay: false,
+		isStripeEceEnabled: false,
```
I figured it'd make sense to initialize the context with ECE disabled since that's the default for now. It may make sense to flip this to `true` if/when we start using ECE by default? I don't think it really matters though.

cesarcosta99 replied:
Yeah, I don't think it really matters. WooPay is still `false`, for example, so it should be totally fine setting it as `false` by default.

reykjalin commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -101,87 +149,78 @@ const PaymentRequestButtonPreview = () => {
+	} else if ( window.location.protocol !== 'https:' ) {
+		return <PreviewRequirementsNotice />;
+	} else if ( isStripeEceEnabled ) {
+		return (
+			<ButtonPreviewWrapper theme={ theme }>
+				<ExpressCheckoutPreviewComponent
+					stripe={ stripe }
+					buttonType={ buttonType }
+					theme={ theme }
+					height={
+						buttonSizeToPxMap[ size ] || buttonSizeToPxMap.medium
+					}
+				/>
+			</ButtonPreviewWrapper>
+		);
+	} else if ( isPaymentRequestEnabled && ! isLoading && paymentRequest ) {
+		<ButtonPreviewWrapper theme={ theme }>
+			<PaymentRequestButtonElement
+				key={ `${ buttonType }-${ theme }-${ size }` }
+				onClick={ ( e ) => {
+					e.preventDefault();
+				} }
+				options={ {
+					paymentRequest: paymentRequest,
+					style: {
+						paymentRequestButton: {
+							type: buttonType,
+							theme: theme,
+							height: `${
+								buttonSizeToPxMap[ size ] ||
+								buttonSizeToPxMap.medium
+							}px`,
+						},
+					},
+				} }
+			/>
+		</ButtonPreviewWrapper>;
+	}
+
+	return <PreviewRequirementsNotice />;
```
The priority of the buttons here is:

1. WooPay button if WooPay is enabled;
2. Info notice if page is not served over HTTPS (ECE and PRBs both require HTTPS to be rendered);
3. ECE buttons (+ WooPay button if enabled);
4. PRBs (+ WooPay button if enabled); and finally
5. Info notice (no buttons could be rendered).

This is how I understood the mass of ternary statements that I refactored, and it makes sense to me to have the list be prioritized like this, but more than happy to have this double checked.

reykjalin commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -67,6 +69,42 @@ const buttonSizeToPxMap = {
+		<div className="payment-method-settings__preview" data-theme={ theme }>
+			{ children }
+		</div>
+		<BrowserHelpText />
```
If we decide to show both button previews when available we need to add the Safari notice only when using Chrome, and never show the notice in Safari, but only when ECE is enabled.

cesarcosta99 replied:
Also related to the inline notice, and now that we're displaying all buttons when available, should we remove the `BrowserHelpText` now that we have this new notice in place? We may display it just on Chrome, however, we may want to hide it when the inline notice is being shown.

**On Brave/Chrome with an insecure connection**

<img width="662" alt="Screenshot 2024-07-08 at 18 27 12" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/42baeba7-ab6f-404d-91fa-a21423b09fe6">

**On Chrome**

<img width="659" alt="Screenshot 2024-07-08 at 18 21 03" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/97a353a3-4435-4747-a43d-846639ad64d5">

**On Safari**

<img width="666" alt="Screenshot 2024-07-08 at 18 24 30" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/ed27c741-06ac-4e22-9d08-5042a879cf26">

reykjalin replied:
Ah, this is a great callout! @nikkivias @pierorocca what do you think here?

elizaan36 replied:
> should we remove the BrowserHelpText now that we have this new notice in place?

Yes, this helper text seems redundant with the notice. I think it could be removed since the notice will provide enough info.

reykjalin replied:
Removed the help text in 4a7cdbf92876153d21a706716f489e836cf03f47.

pierorocca replied:
Gotcha. I missed that you meant the https notice is redundant to the browser notice. Looks good.

Commit:
add missing return statement

Commit:
adapt tests to new notices

Commit:
fix remaining failing tests

cesarcosta99 commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -0,0 +1,107 @@
+
+	const applePayType = buttonType === 'default' ? 'plain' : buttonType;
+
+	let shouldOnlyShowApplePay = false;
+	if ( window.ApplePaySession && window.ApplePaySession.canMakePayments() ) {
+		shouldOnlyShowApplePay = true;
+	}
```
We could use a `useMemo` here to cache this value and avoid multiple calls.

reykjalin replied:
Fixed in 951cd42065acadbd68efe384c5c687cd145f8f22. Great addition! 👏

cesarcosta99 commented on client/express-checkout/blocks/components/express-checkout-preview.js:
```diff
@@ -0,0 +1,107 @@
+		}
+	};
+
+	const googlePayType = buttonType === 'default' ? 'plain' : buttonType;
+
+	const applePayType = buttonType === 'default' ? 'plain' : buttonType;
```
Since we're not distinguishing Google Pay and Apple Pay buttons in this component, should we unify this into a single variable?

reykjalin replied:
Fixed in 3f1e2244815d9564324d2d8d13d0cfb49cb2d282

cesarcosta99 commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -101,87 +149,80 @@ const PaymentRequestButtonPreview = () => {
+		);
+	}
+
+	if ( isWooPayEnabled ) {
```
I'm wondering about why we need to disable WooPay in order to test this? I guess we could handle the payment method separately and be able to preview both ECE and WooPay in their respective pages? Can you shed some light into this?

reykjalin replied:
That's what the previous implementation looked like, so I just copied that in my refactor 🙂 

I _think_ it has something to do with being able to reuse the component for both the WooPay and Express Checkout customization pages, but I'll admit I'm not entirely sure.

I just noticed that the WooPay preview actually looks really broken from the WooPay button settings right now, but it's like that in `develop` as well so I don't think it's a regression in this PR.

reykjalin replied:
>I just noticed that the WooPay preview actually looks really broken from the WooPay button settings right now, but it's like that in develop as well so I don't think it's a regression in this PR.

Turns out this was because I was messing with the settings in the DB, oops. You can ignore that part 😛 

> why we need to disable WooPay in order to test this?

You were right to call this out! It's supposed to show the WooPay button and express checkout buttons side-by-side, I just made a mistake during the refactor ☺️  Nice catch!

reykjalin replied:
Fixed in faa98c9816ee7a83217fce58306b58590487c9cd <3

cesarcosta99 reviewed (changes requested):
I left some comments but this is looking nice and it's testing well 👍 

- Make sure an info notice is displayed when no button can be rendered ✅ 
- Test button settings ⚠️ 
    - I tested all combinations possible for the Apple Pay button ✅ 
    - I tested all combinations possible for the Google Pay button and mostly worked fine. However, there might be a problem when the CTA is "Buy with", it looks like this is the only CTA that's not rendering accordingly. See images below.

<table>
<tr>
<th>Only icon</th>
<th>Buy with</th>
<th>Donate with</th>
<th>Book with</th>
</tr>
<tr>
<td>
<img width="594" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/c232db6b-93fd-4681-8a00-40ea123fbe63">
</td>
<td>
<img width="598" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/02274cb6-a931-44ff-8094-a088219a3a69">
</td>
<td>
<img width="590" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/59a94b7c-17ca-41ac-9646-7644b5637d1b">
</td>
<td>
<img width="590" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/7a65cd71-f5a3-480e-9c90-7b11e4b13212">
</td>
</tr>
</table>

reykjalin commented:
> there might be a problem when the CTA is "Buy with", it looks like this is the only CTA that's not rendering accordingly. See images below.

For whatever reason the `Buy with` text just isn't displayed on the GPay button 🤷‍♂️  If you test Apple Pay you'll see that the CTA is fine, but it won't render on the GPay button and [Stripe uses the same buttonType prop on both buttons for that CTA](https://docs.stripe.com/js/elements_object/create_express_checkout_element#express_checkout_element_create-options-buttonType) (`'buy'`).

So I'm not sure _why_ it's happening, but I suspect it has something to do with "buy" being the default use of the button so maybe they feel that it's redundant?

**EDIT:** I think this has to do with the GPay button showing the last 4 digits of your credit card? You can see in the new recording of multiple buttons showing in the PR description that the narrower version of the button (no last 4 digits) has the correct CTAs.

reykjalin committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

Commit:
fix express checkout button preview composition

Commit:
use same approach to button type for the preview component

Commit:
use useMemo to memoize the result of ApplePaySession.canMakePayments

pierorocca commented:
> Should we have the preview display both buttons when available or try to always limit to just 1 button?
Safari now supports showing both Google Pay and Apple Pay so we won't always know whether the ECE will show only Google Pay or both Apple Pay or Google Pay.

Merchants often are confused by the preview because it doesn't show all the buttons which they've enabled so they think there's something wrong. Let's show all available buttons.


>The method used to decide whether to show only Apple Pay (via ApplePaySession.canMakePayments()) is not fool-proof. We can't tell whether a customer has a card added to their wallet, in which case the button might not show up, resulting in no button in the preview.

There's actually a way in ECE to force this which we may want enable in order to avoid confusion from non-technical merchant users who don't understand why a button may not show up in the preview or the checkout.

> To allow Apple Pay or Google Pay to appear when they’re not set up, set [paymentMethods.applePay](https://docs.stripe.com/js/elements_object/create_express_checkout_element#express_checkout_element_create-options-paymentMethods-applePay) or [paymentMethods.googlePay]

> (https://docs.stripe.com/js/elements_object/create_express_checkout_element#express_checkout_element_create-options-paymentMethods-applePay) to always. This still won’t force them to appear on unsupported platforms, or when the payment is in an unsupported currency.

**Revised**
Option 1: Set the property to `always` for the preview so that merchants will see both Apple and Google Pay regardless if they've setup each wallet (which is unlikely). Set it to always for shoppers as well.

Option 2: Set the property to `always` for the preview so that merchants will see both Apple and Google Pay regardless if they've setup each wallet (which is unlikely). Do NOT set it to `always` for shoppers. Risk here is that merchants when they test checkout as a user, won't see both buttons. The benefit is that shoppers will only see buttons for which they've setup the wallet.

@nikkivias what's your preference between the two options?

pierorocca commented:
@reykjalin there's an open P2 thread on the button alignment in cases where there's overflow to a 2nd row. The feedback is the design guideline for overflow is not balanced when there's only 1 button on the overflow row. As a revision, design will consider centering the button. The attached multiple express buttons preview above looks unbalanced and doesn't match the orientation on checkout. @nikkivias could you provide guidance here please?

The preview here if possible should follow how it will be rendered on the checkout which in the near future may vary between blocks and shortcode.

Commit:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

Commit:
fix invalid 0 radius for express checkout buttons

Commit:
support border radius in ece preview settings

Commit:
fix conditional rendering of ece and prb previews

reykjalin commented:
> Merchants often are confused by the preview because it doesn't show all the buttons which they've enabled so they think there's something wrong. Let's show all available buttons.

Done!

> Option 1: Set the property to always for the preview so that merchants will see both Apple and Google Pay regardless if they've setup each wallet (which is unlikely). Set it to always for shoppers as well.

This is how ECE is currently implemented.

> Option 2: […] Do NOT set it to always for shoppers. Risk here is that merchants when they test checkout as a user, won't see both buttons. The benefit is that shoppers will only see buttons for which they've setup the wallet.

Hmmm, we might be able to use the `auto` option to achieve this without any significant effort.

> The attached multiple express buttons preview above looks unbalanced and doesn't match the orientation on checkout. @nikkivias could you provide guidance here please?
>
> The preview here if possible should follow how it will be rendered on the checkout which in the near future may vary between blocks and shortcode.

Well, none of the locations render the buttons the same way. The preview renders the buttons the same way the Blocks Checkout does it:

| | Shortcode | Blocks |
|:---:|:---:|:---:|
| Checkout | <img width="1124" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/5b3cee5a-4f22-4e0f-8e94-93afb1a59d0f"> | <img width="817" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/efd8fd47-676e-4b06-8733-dadd37d33d06"> |
| Cart | <img width="595" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/019a25d8-f948-4ffd-ba1c-d3ec8b5ce027"> | <img width="595" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/df5594f8-7a01-4445-9404-feb29f83cb3b"> |

I can open a separate PR to try to make the Cart and Shortcode checkout pages look the same, but I don't think we can control the Blocks Checkout any more than we do now.

I can also fix the preview layout so it's just 1 column, but then it won't look like the shortcode version:

<img width="450" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/57059b0e-4928-4c45-a1d2-9f9931fb586e">

Commit:
show all available ECE buttons in settins preview

reykjalin commented:
@cesarcosta99 Since #9010 was merged there's a bit of additional code to review, but it should be minimal. I've also addressed all of your concerns, so we should be looking good for another round of reviews 🎉

pierorocca commented:
> I can open a separate PR to try to make the Cart and Shortcode checkout pages look the same, but I don't think we can control the Blocks Checkout any more than we do now.
I can also fix the preview layout so it's just 1 column, but then it won't look like the shortcode version:

If we change anything, we'd want the preview to represent the shopper experience as much as possible. 

As a compromise, does it make sense to have the preview follow the shortcode checkout layout? Then as Blocks checkout continues to gain traction, merchants can rely on FSE/Blocks editor to preview and adjust settings.

The Blocks version of ECE looks awful compared to PRB in both the cart and checkout. We can't go live with that.
@bborman22 @c-shultz @senadir @alexflorisca could you all weigh in here and help sort of the button layout in the appropriate PR?

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/266ac120-a542-4a38-9ee2-662b149299cf)

bborman22 commented:
> The Blocks version of ECE looks awful compared to PRB in both the cart and checkout.

By this do you mean because of the grid layout? If so, registering the buttons individually in #9047 should address the layout. Is there another visual concern with the blocks version though?

pierorocca commented:
> The Blocks version of ECE looks awful compared to PRB in both the cart and checkout.

>By this do you mean because of the grid layout? If so, registering the buttons individually in https://github.com/Automattic/woocommerce-payments/pull/9047 should address the layout. Is there another visual concern with the blocks version though?

Looking at the included images @bborman22:

1.  The huge gap between the WooPay button and the Apple and Google buttons below it.
2. Calling out the 3 button layout in Blocks Checkout doesn't adhere to design's guidelines AND chatting with Daniel he's agreed the design guideline itself looks unbalanced and that the 3rd button on the 2nd row might be better centered. Looking for @alexflorisca's guidance on 2.

Does #9047 fix 1.? Note it's hard to tell as #9047 and #9045 don't describe the problem being solved.

A follow-up question to #9047, is PayPal registrering each button as individual payment methods?

Commit:
change button layout in preview to resemble shortcode pages

reykjalin commented:
> As a compromise, does it make sense to have the preview follow the shortcode checkout layout? Then as Blocks checkout continues to gain traction, merchants can rely on FSE/Blocks editor to preview and adjust settings.

I think that's good compromise for now. Here's what that looks like:

| Desktop view | Mobile view |
|:---:|:---:|
| <img width="675" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/9596786d-54dc-4ff3-86cc-ad3a954e10b7"> | <img width="484" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/70e1c77f-b581-4713-a65e-4f439a53cafd"> |

I've pushed this in d8f1ee0f87fdcedc29374419cfe8adc7415bb6be.

> Does https://github.com/Automattic/woocommerce-payments/pull/9047 fix [The huge gap between the WooPay button and the Apple and Google buttons below it.]?

I don't think it will. I remember trying to look into this and [I couldn't figure out where the extra spacing is coming from](https://github.com/Automattic/woocommerce-payments/issues/8951#issuecomment-2183475932).

> is PayPal registrering each button as individual payment methods?

I don't know, we'd have to check the plugin code. But if they properly arrange themselves in the Blocks Express Checkout container then they've either loaded each button separately, or they've put in some effort to match the breakpoints in the Checkout block precisely.

reykjalin commented:
As for the design issues I'm going to open a separate issue for that. I think it'll be better to figure out what sort of layout we're going to implement and then update the preview to follow that, rather than doing something now with the preview that may not end up aligning with the design we arrive at.

reykjalin commented:
@pierorocca @bborman22  I've opened https://github.com/Automattic/woocommerce-payments/issues/9066#issue-2396443411 to track the remaining design issues we need to resolve.

bborman22 commented:
> A follow-up question to https://github.com/Automattic/woocommerce-payments/pull/9047, is PayPal registrering each button as individual payment methods?

I'm not able to get the Venmo button to load on cart block and checkout block (though it loads on product and shortcodes). But looking into it, they do appear to register each button individually with Blocks. In my case with Venmo not loading that also creates a gap similar to what we were witnessing with ECE and exploring ways to address it.

![afjueI.png](https://github.com/Automattic/woocommerce-payments/assets/68524302/1cb68cdf-4e1a-474e-909f-80f4446f7fbf)

cesarcosta99 commented on client/settings/express-checkout-settings/payment-request-button-preview.js:
```diff
@@ -103,88 +152,84 @@ const PaymentRequestButtonPreview = () => {
+			<ButtonPreviewWrapper theme={ theme }>
+				{ woopayPreview }
+				{ /* We never want to show both ECE and PRB previews at the same time. */ }
+				{ ( window.location.protocol === 'https:' &&
```
This check is preventing the [inline notice](https://github.com/Automattic/woocommerce-payments/pull/9028/files#diff-8368d420c86a67fc8fc3bceafa24455eeaf836e4dca6e0c1de3da227cae67604R97) from being displayed in case the settings is accessed through an insecure connection. I believe we can remove it in favor of the notice.

One thing to note is that the inline notice may also be displayed when user is customizing the WooPay button. I guess this is expected (and desired)? cc @pierorocca

reykjalin replied:
Ah, sorry, I forgot to update the testing instructions here. You need to remove this check for the `Make sure an info notice is displayed when no button can be rendered` test to work.

The check is there to give the merchant better information on why the button might not be rendering. The test is to check the more generic fallback in the ECE preview component that could, in theory, be triggered by other failures. Trying to open the page on HTTP is just the only way I know to force the buttons to fail rendering 😅 

So yeah, maybe you can try re-running that test after removing the `window.location.protocol === 'https:'` check?

reykjalin replied:
I've also updated the testing instructions 👍

cesarcosta99 replied:
I understand what you're trying to achieve here, and the test case was correct previously. What I'm calling out is that the notice will never be displayed if the user is on an insecure connection because the current logic doesn't allow that. What do you think about moving that check inside the component to prevent the buttons to render and enable the notice in such cases, maybe to the [onReady callback](https://github.com/Automattic/woocommerce-payments/pull/9028/files#diff-8368d420c86a67fc8fc3bceafa24455eeaf836e4dca6e0c1de3da227cae67604R71-R77)?

reykjalin replied:
It actually does show up in Firefox:

<img width="706" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/4e79d5b5-69fd-4679-80f6-a70b68927579">

Should've just made that the test case, don't know why I didn't think of that before.

But yeah, it is absolutely necessary when the merchant tries to look at the preview in a browser that isn't configured correctly or doesn't support Apple Pay or Google Pay.

reykjalin replied:
Moving the `https` check into the component would cause the notice not to appear when ECE is disabled, because the notice catches the `HTTPS` condition for both ECE and PRBs. That's why I don't think the `HTTPS` check should be in the ECE preview component itself.

reykjalin replied:
> One thing to note is that the inline notice may also be displayed when user is customizing the WooPay button. I guess this is expected (and desired)? cc @pierorocca

I _think_ this is desired since the WooPay button appears in the GPay/APay customization menu? It makes sense to me that the same should apply in the WooPay customization menu so you can see what your checkout will actually look like.

pierorocca replied:
> I think this is desired since the WooPay button appears in the GPay/APay customization menu? It makes sense to me that the same should apply in the WooPay customization menu so you can see what your checkout will actually look like.

That logic makes sense to me @reykjalin. When we get to refreshing the payment settings and consolidating where we can, this won't be as confusing of an experience.

cesarcosta99 replied:
>maybe you can try re-running that test after removing the window.location.protocol === 'https:' check?

I don't think removing that check to test this is the right call because we can simulate it being `true` or `false` without removing it.

>It actually does show up in Firefox:

I don't see the notice in Firefox when testing through a non-HTTPS connection.

<img width="1189" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/15b83d3f-27b9-4c7e-a3d9-93c79cc2f3fe">


>Moving the https check into the component would cause the notice not to appear when ECE is disabled, because the notice catches the HTTPS condition for both ECE and PRBs.

That's true. Then, we can tweak the logic to pass its value over to the component instead of conditionally render it and prevent the notice from rendering. The `https` check has been introduced in this PR, so I'm confident we can restructure it to please all our needs 🙂 Unless we have a strong reason not to do so?

I really would like this test case to be successful because I'm sure there will be folks accessing this page via HTTP and the fact that they don't see the ECE/PRB button neither a notice may cause confusion.

reykjalin replied:
> I don't see the notice in Firefox when testing through a non-HTTPS connection.

Ah, this was actually a bug! Or sort of intended? Basically to only show the WooPay button when that's the only button available. I fixed this in 245091b785d41f0c39b73346bf2b74b8abe5e4b2 though.

<details><summary>A chart of what the various settings now look like:</summary>

| Browser | HTTPS? | WooPay? | APay/GPay? | Screenshot |
|:---:|:---:|:---:|:---:|:---:|
| Firefox | ⛔️ | ⛔️ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/ff830df1-834b-4105-8bf5-70a28fcd9e83"> |
| Firefox | ⛔️ | ✅ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/1e9fceba-63fb-49d5-b48d-47f48ac2ec50"> |
| Firefox | ⛔️ | ⛔️ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/b72e02f5-c344-4e34-80cd-f1414fcb5604"> |
| Firefox | ⛔️ | ✅ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/b175c6b2-c98c-48bd-b1d7-aa454f5c7355"> |
| Firefox | ✅ | ⛔️ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/a8487dd9-ee7e-4d27-b2ab-9b83ef1f69be"> |
| Firefox | ✅ | ✅ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/16cf4e89-8ac1-4cac-8900-fee3e84ca49c"> |
| Firefox | ✅ | ⛔️ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/ae4db014-a326-4076-9a07-e593f224a7a6"> |
| Firefox | ✅ | ✅ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/ea49462c-d2b7-4db8-b1ed-765b6a6684f1"> |
| Chrome | ⛔️ | ⛔️ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/1468e983-520b-49a4-9db2-7067e101cf10"> |
| Chrome | ⛔️ | ✅ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/3fa92e35-a471-4fa0-a7fc-1b40d0674e74"> |
| Chrome | ⛔️ | ⛔️ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/f62b616a-445f-4d6f-a309-c39404b910b9"> |
| Chrome | ⛔️ | ✅ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/38b533e7-8409-43b8-b751-60f46fd69c72"> |
| Chrome | ✅ | ⛔️ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/651c66f5-b759-490b-b55e-8378d428b23d"> |
| Chrome | ✅ | ✅ | ⛔️ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/ddc6914e-41f9-4c55-83b8-994348232466"> |
| Chrome | ✅ | ⛔️ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/0ef22a89-2e11-4cb7-b75e-ad1d9d851ac6"> |
| Chrome | ✅ | ✅ | ✅ | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/8df73fa5-f80b-45ac-bd1f-2048df95044d"> |

</details>

> Then, we can tweak the logic to pass its value over to the component instead of conditionally render it and prevent the notice from rendering. The https check has been introduced in this PR, so I'm confident we can restructure it to please all our needs 🙂 Unless we have a strong reason not to do so?

We _could_ but I don't really like it because we'd need to either duplicate the notice, or import it from another file. Or pass it in as some sort of `fallbackNotice` prop on both the PRB and ECE preview components.

The nice thing about the if-statements in the `<PaymentRequestButtonPreview />` parent component is that it's easy to find all the components it renders instead of drilling down into each component. I think it feels better, architecturally speaking, to keep the requirements info notice in the `<PaymentRequestButtonPreview />` component instead of passing more state down to the `<ExpressCheckoutPreviewComponent />` and `<PaymentRequestButtonElement />` components.

> I really would like this test case to be successful because I'm sure there will be folks accessing this page via HTTP and the fact that they don't see the ECE/PRB button neither a notice may cause confusion.

This should be addressed by the latest commit 👍

cesarcosta99 replied:
That looks great 🎉 Thank you for addressing this, Kris!

cesarcosta99 reviewed (changes requested):
Thanks for addressing the requested changes 🎉 The code looks good, but a test case requires attention:

- Make sure an info notice is displayed when no button can be rendered ⚠️ 
    - Check comments below.
- Test button settings ✅ 
    - All combinations possible for the Apple Pay button
    - All combinations possible for the Google Pay button 

>I think this has to do with the GPay button showing the last 4 digits of your credit card? You can see in the new recording of multiple buttons showing in the PR description that the narrower version of the button (no last 4 digits) has the correct CTAs.

I think you're right 👍 When I tested it on Safari (because now Google Pay is displayed there as well), where I'm not logged on my Google account, the "Buy with" CTA is being displayed.

pierorocca commented:
> Also related to the inline notice, and now that we're displaying all buttons when available, should we remove the BrowserHelpText now that we have this new notice in place? We may display it just on Chrome, however, we may want to hide it when the inline notice is being shown.

It's a good callout @cesarcosta99. Apple announced at WWDC24 that they'll bring Apple Pay to all browsers on Mac OS this year. I don't know the timing but I'd expect Stripe to follow suit and support it. That would close a big chunk of [non-supported browsers.](https://docs.stripe.com/elements/express-checkout-element#supported-browsers)

Based on the Stripe supported browser list, the two most relevant scenarios where a button fails to display are:

1. Displaying Apple Pay when Safari is not used on desktop (Apple is adding support soon)
2. Displaying Google Pay when Firefox is used on desktop.

For 1, display "To preview the Apple Pay button, view this page in the Safari browser."
For 2, display "To preview the Apple Pay and Google Pay buttons, view this page in the Safari and Chrome browsers."

Is this an appropriate option?

pierorocca commented:
> In my case with Venmo not loading that also creates a gap similar to what we were witnessing with ECE and exploring ways to address it.

Thanks @bborman22. Do you think this is WooPayments caused or a general issue with the checkout block?

elizaan36 commented:
> what sort of layout we're going to implement and then update the preview to follow that

@pierorocca @reykjalin 👋 The buttons should wrap and align left on the 2nd line. Not sure why they're aligning to the right in the [screenshot here](https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2211971909) for blocks, this seems like a bug.

alexflorisca commented:
> [The huge gap between the WooPay button and the Apple and Google buttons below it.]?

I've seen this too in some of my testing. I don't know if this comes from WooPayments or not yet but it's worth looking into. Tends to happen in Safari and Firefox

> Calling out the 3 button layout in Blocks Checkout doesn't adhere to design's guidelines AND chatting with Daniel he's agreed the design guideline itself looks unbalanced and that the 3rd button on the 2nd row might be better centered. Looking for @alexflorisca's guidance on 2.

I've just crated a [PR](https://github.com/woocommerce/woocommerce/pull/49304) to address the layout problem. Each button will have a minimum width of 240px, all buttons displayed equally widths, and a max of 2 columns. It'll look something like this

https://github.com/Automattic/woocommerce-payments/assets/3966773/c0249cb8-5ac8-4863-bfd6-30139bb8acc2

For me the buttons render ok with ECE when they are rendered individually.

asumaran committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

Commit:
update info and error notice text to be more concise

Commit:
fix bug where no info notice is shown when all express checkout options are disabled

Commit:
remove browser help text component

Commit:
fix failing tests

reykjalin commented:
> For 1, display "To preview the Apple Pay button, view this page in the Safari browser."
> For 2, display "To preview the Apple Pay and Google Pay buttons, view this page in the Safari and Chrome browsers."
>
> Is this an appropriate option?

@pierorocca We're removing the help text: https://github.com/Automattic/woocommerce-payments/pull/9028#discussion_r1669612375

> The buttons should wrap and align left on the 2nd line. Not sure why they're aligning to the right in the https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2211971909 for blocks, this seems like a bug.

@elizaan36 It's a bug, but at the same time unavoidable _in the current implementation_. The way the new ECE components work is that multiple buttons are rendered in one container. The Checkout block has different breakpoints for wrapping content than Stripe's ECE container, so you end up with a weird layout.

We're going to address this with #9045 and render multiple ECE containers so each ECE button will be rendered individually.

pierorocca commented:
> > For 1, display "To preview the Apple Pay button, view this page in the Safari browser."
> > For 2, display "To preview the Apple Pay and Google Pay buttons, view this page in the Safari and Chrome browsers."
> > Is this an appropriate option?
> 
> @pierorocca We're removing the help text: [#9028 (comment)](https://github.com/Automattic/woocommerce-payments/pull/9028#discussion_r1669612375)

That works when the insecure notice is presented. When it's a secure connection and there's no notice, as hightlighted here https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2216172740 there are still 2 cases where the buttons shown don't match what the merchant is expecting to see. Are you planning to use the same inline notice with different copy? Or am I missing something?

reykjalin commented:
> When it's a secure connection and there's no notice, as hightlighted here https://github.com/Automattic/woocommerce-payments/pull/9028#issuecomment-2216172740 there are still 2 cases where the buttons shown don't match what the merchant is expecting to see.

@pierorocca We always show a notice when a button isn't loaded, and I'd argue if the merchant is in Chrome they should only expect to see the Google Pay button. See the screenshots:

| Browser | Screenshot |
|:---:|:---:|
| Chrome (http, logged in, logged out) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/02f4610e-57e2-4767-8c63-60e191fd9fc3"> |
| Chrome (https, logged into GPay) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/7da3d49b-d6e9-437c-aaf3-388a3947eef4"> |
| Chrome (https, logged out of GPay) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/7716aa6a-4eeb-402e-9905-aec12d6c32b0"> |
| Safari (http) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/b2ba0e80-f393-42b8-ae61-0361674d5a5f"> |
| Safari (https) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/af0f24c6-645d-4db5-9831-7491744cffbd"> |
| Firefox (http) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/5cf28b55-a96f-4a6c-b9ee-115e037acc0e"> |
| Firefox (https) | <img width="716" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/80dfbffa-4815-4c4e-bafc-3dd54148271e"> |

Commit:
show info notice only when buttons are enabled and https is not detected

reykjalin committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview

pierorocca commented:
That works. There might be some merchant confusion in the short term based on support tickets I've seen in the past. As Apple and Stripe add non Safari support for Apple Pay Later this year that confusion will become a non-issue.

The https version for Firefox is confusing to me but I'm fine with it to simplify the implementation. Firefox has a market share size too small to make it relevant.

I appreciate the visuals, they helped a lot.

cesarcosta99 reviewed (approved):
Thanks for all your hard work here! Code LGTM and all tests passed 🚀 

- Make sure an info notice is displayed when no button can be rendered ✅ 
    - Tested with:
        - Firefox, Chrome and Safari.
        - HTTP and HTTPS.
        - WooPay and Apple Pay and Google Pay enabled and disabled.
- Test button settings ✅
    - All combinations possible for the Apple Pay button
    - All combinations possible for the Google Pay button

reykjalin committed:
Merge branch 'develop' into add/9007-use-ece-for-express-button-preview
